src/optimisation/Optimisation.jl

```diff
@@ -139,186 +154,364 @@ function (f::OptimLogDensity)(F, G, z)
-    ctx = OptimizationContext(DynamicPPL.DefaultContext())
-    obj = OptimLogDensity(model, ctx)
+"""
+    estimate_mode(
+        model::DynamicPPL.Model,
+        estimator::ModeEstimator,
+        [initial_params::Union{AbstractVector,Nothing}],
+        [solver];
+        <keyword arguments>
+    )
+
+Find the mode of the probability distribution of a model.
+
+Under the hood this function calls `Optimization.solve`.
+
+# Arguments
+- `model::DynamicPPL.Model`: The model for which to estimate the mode.
+- `estimator::ModeEstimator`: Can be either `MLE()` for maximum likelihood estimation or
+    `MAP()` for maximum a posteriori estimation.
+- `initial_params::Union{AbstractVector,Nothing}`: Initial value for the optimization.
+    Optional, unless non-box constraints are specified. If `nothing` or omitted it is
+    generated by either sampling from the prior distribution or uniformly from the box
+    constraints, if any.
+- `solver`. The optimization algorithm to use. Optional. Can be any solver recognised by
+    Optimization.jl. If `nothing` or omitted a default solver is used: LBFGS, or IPNewton if
+    non-box constraints are present.
+
+# Keyword arguments
+- `adtype::AbstractADType=AutoForwardDiff()`: The automatic differentiation type to use.
+- Keyword arguments `lb`, `ub`, `cons`, `lcons`, and `ucons` define constraints for the
+    optimization problem. Please see [`ModeEstimationConstraints`](@ref) for more details.
+- Any extra keyword arguments are passed to `Optimization.solve`.
+"""
+function estimate_mode(
+    model::DynamicPPL.Model,
+    estimator::ModeEstimator,
+    initial_params::Union{AbstractVector,Nothing},
+    solver;
+    adtype=ADTypes.AutoForwardDiff(),
+    cons=nothing,
+    lcons=nothing,
+    ucons=nothing,
+    lb=nothing,
+    ub=nothing,
+    kwargs...
+)
+    constraints = ModeEstimationConstraints(lb, ub, cons, lcons, ucons)
+    initial_params = generate_initial_params(model, initial_params, constraints)
+    if solver === nothing
+        solver = default_solver(constraints)
+    end
 
-    obj = transform!!(obj)
-    init = Init(obj.varinfo, model, constrained_space{false}())
-    t = ParameterTransform(obj.varinfo, model, constrained_space{true}())
+    # Create an OptimLogDensity object that can be used to evaluate the objective function,
+    # i.e. the negative log density. Set its VarInfo to the initial parameters.
+    log_density = let
+        inner_context = if estimator isa MAP
+            DynamicPPL.DefaultContext()
+        else
+            DynamicPPL.LikelihoodContext()
+        end
+        ctx = OptimizationContext(inner_context)
+        ld = OptimLogDensity(model, ctx)
+        Accessors.@set ld.varinfo = DynamicPPL.unflatten(ld.varinfo, initial_params)
+    end
 
-    return (obj=obj, init = init, transform=t)
-end
+    # TODO(mhauru) We currently couple together the questions of whether the user specified
```

Comment:
Do you want the keyword arg now, even before we implement that? Would have to start throwing errors if the user sets it to `false` without giving explicit bounds, or conversely leaves it `true` when providing explicit bounds. I think that could be quite confusing, even if it would mean that our interface could be a bit more stable in the future.

Reply:
No, we introduce the kwarg with the functionality, not before :+1: